src/xenia/kernel/xboxkrnl/xboxkrnl_ob.cc

```diff
@@ -78,70 +78,40 @@ DECLARE_XBOXKRNL_EXPORT1(ObLookupThreadByThreadId, kNone, kImplemented);
 dword_result_t ObReferenceObjectByHandle(dword_t handle,
                                          dword_t object_type_ptr,
                                          lpdword_t out_object_ptr) {
-  X_STATUS result = X_STATUS_SUCCESS;
+  const std::pair<XObject::Type, uint32_t> obj_type_match[] = {
+      {XObject::kTypeEvent, 0xD00EBEEF},
+      {XObject::kTypeSemaphore, 0xD017BEEF},
+      {XObject::kTypeThread, 0xD01BBEEF}};
 
   auto object = kernel_state()->object_table()->LookupObject<XObject>(handle);
-  if (object) {
-    // TODO(benvanik): verify type with object_type_ptr
-
-    // TODO(benvanik): get native value, if supported.
-    uint32_t native_ptr;
-    switch (object_type_ptr) {
-      case 0x00000000: {  // whatever?
-        switch (object->type()) {
-          case XObject::kTypeEvent: {
-            assert(object->type() == XObject::kTypeEvent);
-            native_ptr = object->guest_object();
-            assert_not_zero(native_ptr);
-          } break;
-          case XObject::kTypeSemaphore: {
-            assert(object->type() == XObject::kTypeSemaphore);
-            native_ptr = object->guest_object();
-            assert_not_zero(native_ptr);
-          } break;
-          case XObject::kTypeThread: {
-            assert(object->type() == XObject::kTypeThread);
-            native_ptr = object->guest_object();
-            assert_not_zero(native_ptr);
-          } break;
-          default: {
-            assert_unhandled_case(object->type());
-            native_ptr = 0xDEADF00D;
-          } break;
-        }
-      } break;
-      case 0xD00EBEEF: {  // ExEventObjectType
-        assert(object->type() == XObject::kTypeEvent);
-        native_ptr = object->guest_object();
-        assert_not_zero(native_ptr);
-      } break;
-      case 0xD017BEEF: {  // ExSemaphoreObjectType
-        assert(object->type() == XObject::kTypeSemaphore);
-        native_ptr = object->guest_object();
-        assert_not_zero(native_ptr);
-      } break;
-      case 0xD01BBEEF: {  // ExThreadObjectType
-        assert(object->type() == XObject::kTypeThread);
-        native_ptr = object->guest_object();
-        assert_not_zero(native_ptr);
-      } break;
-      default: {
-        assert_unhandled_case(object_type_ptr);
-        native_ptr = 0xDEADF00D;
-      } break;
-    }
 
-    // Caller takes the reference.
-    // It's released in ObDereferenceObject.
-    object->RetainHandle();
-    if (out_object_ptr.guest_address()) {
-      *out_object_ptr = native_ptr;
+  if (!object) {
+    return X_STATUS_INVALID_HANDLE;
+  }
+
+  uint32_t native_ptr = object->guest_object();
+
+  auto obj_type = std::find_if(
```

Comment:
Maybe make `obj_type_match` an std::map instead to make it more obvious how this structure is intended to be used.

Reply:
I suggested against it because:
1) It has constant contents and size known at compile time
2) For such small samples binary search isn't faster than linear but has more code overhead

Reply:
AFAIK you can declare an `std::map` the same way. You're right about the search potentially being slower, though it likely does not matter here unless a game hammers on this function.
We can use `std::unordered_map` instead because order does not matter.

Reply:
Indeed it's possible - so I guess it's just a matter of preference.